### ISSUE TITLE
Deadlock on stack overflow

### ISSUE THREAD

**Issue by noahfalk** [dotnet/runtime#63370]
### Description

In some cases when a .NET app triggers a stack overflow we deadlock while trying to do our error handling HandleFatalStackOverflow. For many services there are automatic mechanisms to relaunch an app that crashes, but apps which hang are harder to detect and recover so it is greatly prefered for the app to exit. 

The underlying cause appears to be two conflicting assumptions:
1. Code in the standard exception handling path assumes it is safe to invoke the CRT heap allocator while in co-op mode. That allocator has a lock which could block on another thread.
2. Code in the fatal error handling path assumes that we can reliably do a mode transition after an SO has occured. 

The most straightforward solution I see so far would be modifying HandleFatalStackOverflow to do a transition from co-op->preemptive before it invokes DoFavor and then it would never transition back because the function doesn't return. This would avoid having the Crst inside DoFavor trying to toggle the GC mode.

### Reproduction Steps

I don't have a repro but we do have stack traces from a dump that do a pretty good job of illustrating the problem:

This thread is in cooperative mode and won’t exit cooperative mode until it is able to complete it’s CRT heap allocation
```
  56  Id: 1b320.1bfc4 Suspend: 0 Teb: 000000a6`8afe1000 Unfrozen ""
# Child-SP          RetAddr               Call Site
00 000000a6`93a359f8 00007ffe`aa2ddc7d     ntdll!ZwWaitForAlertByThreadId+0x14 [minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 4019] 
01 000000a6`93a35a00 00007ffe`aa2dd9a1     ntdll!RtlpWaitOnAddressWithTimeout+0x81 [minkernel\ntos\rtl\waitaddr.c @ 851] 
02 (Inline Function) --------`--------     ntdll!RtlpWaitOnAddress+0xb7 [minkernel\ntos\rtl\waitaddr.c @ 1094] 
03 000000a6`93a35a30 00007ffe`aa2fad6c     ntdll!RtlpWaitOnCriticalSection+0x1a1 [minkernel\ntos\rtl\resource.c @ 1610] 
04 000000a6`93a35b30 00007ffe`aa2fabd2     ntdll!RtlpEnterCriticalSectionContended+0x18c [minkernel\ntos\rtl\resource.c @ 2317] 
05 000000a6`93a35b90 00007ffe`aa2eef62     ntdll!RtlEnterCriticalSection+0x42 [minkernel\ntos\rtl\resource.c @ 1923] 
06 (Inline Function) --------`--------     ntdll!RtlAcquireLockRoutine+0x21 [minkernel\ntos\rtl\heappriv.h @ 383] 
07 000000a6`93a35bc0 00007ffe`aa2ef2a4     ntdll!RtlpAllocateUserBlockFromHeap+0x56 [minkernel\ntos\rtl\heaplowf.c @ 557] 
08 000000a6`93a35c60 00007ffe`aa2e9ffd     ntdll!RtlpAllocateUserBlock+0x118 [minkernel\ntos\rtl\heaplowf.c @ 810] 
09 000000a6`93a35cb0 00007ffe`aa2e916a     ntdll!RtlpLowFragHeapAllocFromContext+0x67d [minkernel\ntos\rtl\heaplowf.c @ 4590] 
0a 000000a6`93a35d70 00007ffe`0975ea31     ntdll!RtlpAllocateHeapInternal+0x12a [minkernel\ntos\rtl\heap.c @ 2368] 
0b (Inline Function) --------`--------     coreclr!ClrMalloc+0x1e [D:\workspace\_work\1\s\src\coreclr\src\utilcode\clrhost_nodependencies.cpp @ 258]
0c (Inline Function) --------`--------     coreclr!operator new[]+0x1e [D:\workspace\_work\1\s\src\coreclr\src\utilcode\clrhost_nodependencies.cpp @ 382] 
0d 000000a6`93a35e70 00007ffe`096fee28     coreclr!StackTraceInfo::AllocateStackTrace+0x31 [D:\workspace\_work\1\s\src\coreclr\src\vm\excep.cpp @ 3302] 
0e 000000a6`93a35ea0 00007ffe`096ff027     coreclr!ExceptionTracker::GetOrCreateTracker+0x1d0 [D:\workspace\_work\1\s\src\coreclr\src\vm\exceptionhandling.cpp @ 3858] 
0f 000000a6`93a35f30 00007ffe`aa36a94f     coreclr!ProcessCLRException+0x167 [D:\workspace\_work\1\s\src\coreclr\src\vm\exceptionhandling.cpp @ 999] 
10 000000a6`93a36160 00007ffe`aa2f61aa     ntdll!RtlpExecuteHandlerForException+0xf [minkernel\ntos\rtl\amd64\xcptmisc.asm @ 132] 
11 000000a6`93a36190 00007ffe`aa2f3331     ntdll!RtlDispatchException+0x25a [minkernel\ntos\rtl\amd64\exdsptch.c @ 655] 
12 000000a6`93a368e0 00007ffe`a78e7d7c     ntdll!RtlRaiseException+0x1f1 [minkernel\ntos\rtl\amd64\raise.c @ 240] 
13 000000a6`93a370c0 00007ffe`09709b46     KERNELBASE!RaiseException+0x6c [minkernel\kernelbase\xcpt.c @ 936] 
14 000000a6`93a371a0 00007ffe`09709619     coreclr!RaiseTheExceptionInternalOnly+0x29a [D:\workspace\_work\1\s\src\coreclr\src\vm\excep.cpp @ 2806] 
15 000000a6`93a372b0 00007ffd`ac391cd0     coreclr!IL_Throw+0xd9 [D:\workspace\_work\1\s\src\coreclr\src\vm\jithelpers.cpp @ 4185] 
```

This thread is in preemptive mode and is blocked waiting for the GC to complete to do a transition back to co-op, it holds the CRT heap lock
```
  54  Id: 1b320.f04c Suspend: 0 Teb: 000000a6`8aec3000 Unfrozen ""
# Child-SP          RetAddr               Call Site
00 000000a6`937316e8 00007ffe`a78d01de     ntdll!ZwWaitForSingleObject+0x14 [minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 251] 
01 000000a6`937316f0 00007ffe`096fcfd0     KERNELBASE!WaitForSingleObjectEx+0x8e [minkernel\kernelbase\synch.c @ 1328] 
02 (Inline Function) --------`--------     coreclr!CLREventWaitHelper2+0x6 [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 376] 
03 000000a6`93731790 00007ffe`096a54a5     coreclr!CLREventWaitHelper+0x20 [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 401] 
04 000000a6`937317f0 00007ffe`097cad0e     coreclr!Thread::RareDisablePreemptiveGC+0x1ad [D:\workspace\_work\1\s\src\coreclr\src\vm\threadsuspend.cpp @ 2387] 
05 (Inline Function) --------`--------     coreclr!Thread::DisablePreemptiveGC+0x13f861 [D:\workspace\_work\1\s\src\coreclr\src\vm\threads.h @ 1974] 
06 000000a6`93731860 00007ffe`099a6430     coreclr!CrstBase::Enter+0x13f8fe [D:\workspace\_work\1\s\src\coreclr\src\vm\crst.cpp @ 333] 
07 (Inline Function) --------`--------     coreclr!CrstBase::AcquireLock+0x8 [D:\workspace\_work\1\s\src\coreclr\src\vm\crst.h @ 187] 
08 (Inline Function) --------`--------     coreclr!CrstBase::CrstHolder::{ctor}+0x10 [D:\workspace\_work\1\s\src\coreclr\src\vm\crst.h @ 382] 
09 000000a6`93731890 00007ffe`0999a5c6     coreclr!DebuggerRCThread::DoFavor+0x3c [D:\workspace\_work\1\s\src\coreclr\src\debug\ee\rcthread.cpp @ 1724] 
0a 000000a6`93731900 00007ffe`0987d994     coreclr!Debugger::RequestFavor+0x46 [D:\workspace\_work\1\s\src\coreclr\src\debug\ee\debugger.cpp @ 2461] 
0b 000000a6`93731930 00007ffe`0987da84     coreclr!EEPolicy::HandleFatalStackOverflow+0x2a0 [D:\workspace\_work\1\s\src\coreclr\src\vm\eepolicy.cpp @ 692] 
0c 000000a6`93731ed0 00007ffe`097ea36c     coreclr!EEPolicy::HandleStackOverflow+0x74 [D:\workspace\_work\1\s\src\coreclr\src\vm\eepolicy.cpp @ 149] 
0d 000000a6`93731f10 00007ffe`aa36a94f     coreclr!ProcessCLRException+0xeb4ac [D:\workspace\_work\1\s\src\coreclr\src\vm\exceptionhandling.cpp @ 893] 
0e 000000a6`93732140 00007ffe`aa2f61aa     ntdll!RtlpExecuteHandlerForException+0xf [minkernel\ntos\rtl\amd64\xcptmisc.asm @ 132] 
0f 000000a6`93732170 00007ffe`aa36993e     ntdll!RtlDispatchException+0x25a [minkernel\ntos\rtl\amd64\exdsptch.c @ 655] 
10 000000a6`937328c0 00007ffe`aa2e908f     ntdll!KiUserExceptionDispatch+0x2e [minkernel\ntos\rtl\amd64\trampoln.asm @ 755] 
11 000000a6`93732fc0 00007ffe`aa2eef74     ntdll!RtlpAllocateHeapInternal+0x4f [minkernel\ntos\rtl\heap.c @ 2257] 
12 000000a6`937330c0 00007ffe`aa2ef2a4     ntdll!RtlpAllocateUserBlockFromHeap+0x68 [minkernel\ntos\rtl\heaplowf.c @ 558]
```

This thread is trying to do a GC and is blocked until all threads exit co-op mode:
```
   3  Id: 1b320.30c54 Suspend: 0 Teb: 000000a6`8aebb000 Unfrozen ""
# Child-SP          RetAddr               Call Site
00 000000a6`8ba7f178 00007ffe`a78d01de     ntdll!ZwWaitForSingleObject+0x14 [minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 251] 
01 000000a6`8ba7f180 00007ffe`096fcfd0     KERNELBASE!WaitForSingleObjectEx+0x8e [minkernel\kernelbase\synch.c @ 1328] 
02 (Inline Function) --------`--------     coreclr!CLREventWaitHelper2+0x6 [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 376] 
03 000000a6`8ba7f220 00007ffe`097010a1     coreclr!CLREventWaitHelper+0x20 [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 401] 
04 (Inline Function) --------`--------     coreclr!CLREventBase::WaitEx+0xd [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 470] 
05 (Inline Function) --------`--------     coreclr!CLREventBase::Wait+0xd [D:\workspace\_work\1\s\src\coreclr\src\vm\synch.cpp @ 416] 
06 000000a6`8ba7f280 00007ffe`09755973     coreclr!ThreadSuspend::SuspendRuntime+0x239 [D:\workspace\_work\1\s\src\coreclr\src\vm\threadsuspend.cpp @ 3859] 
07 000000a6`8ba7f340 00007ffe`096f30a4     coreclr!ThreadSuspend::SuspendEE+0x283 [D:\workspace\_work\1\s\src\coreclr\src\vm\threadsuspend.cpp @ 6144] 
08 (Inline Function) --------`--------     coreclr!GCToEEInterface::SuspendEE+0x26 [D:\workspace\_work\1\s\src\coreclr\src\vm\gcenv.ee.cpp @ 26] 
09 000000a6`8ba7f4e0 00007ffe`099671b8     coreclr!WKS::GCHeap::GarbageCollectGeneration+0x154 [D:\workspace\_work\1\s\src\coreclr\src\gc\gc.cpp @ 37714]
```

In short thread 3 is waiting on thread 56, 56 is waiting on 54, and 54 is waiting on 3. 

### Expected behavior

StackOverflow error handling completes, the thread exits with an unhandled exception and the app is terminated by the OS.

### Actual behavior

StackOverflow error handling hangs, the app isn't terminated.

### Regression?

Not that I am aware of

### Known Workarounds

Not that I am aware of

### Configuration

Windows x64, not sure of the coreclr version but the issue is likely present in many versions
